src/Compilers/CSharp/Test/Emit2/Semantics/ExperimentalAttributeTests.cs

```diff
@@ -0,0 +1,880 @@
+
+        comp.VerifyDiagnostics(
+            // (1,1): warning DiagID1: 'D' is for evaluation purposes only and is subject to change or removal in future updates.
+            // D d = null;
+            Diagnostic("DiagID1", "D").WithArguments("D").WithLocation(1, 1)
+            );
+    }
+
+    [Theory, CombinatorialData]
+    public void OnParameter(bool inSource)
+    {
+        // Ignored on parameters
+        var libSrc = """
+public class C
+{
+    public static void M([System.Diagnostics.CodeAnalysis.Experimental("DiagID1")] int i) { }
+}
+""";
+
+        var src = """
+C.M(42);
+""";
+
+        var comp = inSource
+            ? CreateCompilation(new[] { src, libSrc, experimentalAttributeSrc })
+            : CreateCompilation(src, references: new[] { CreateCompilation(new[] { libSrc, experimentalAttributeSrc }).EmitToImageReference() });
+
+        comp.VerifyDiagnostics();
+    }
+
+    [Theory, CombinatorialData]
+    public void OnReturnValue(bool inSource)
+    {
+        // Ignored on return value
+        var libSrc = """
+public class C
+{
+    [return: System.Diagnostics.CodeAnalysis.Experimental("DiagID1")]
+    public static int M() => 0;
+}
+""";
+
+        var src = """
+_ = C.M();
+""";
+
+        var comp = inSource
+            ? CreateCompilation(new[] { src, libSrc, experimentalAttributeSrc })
+            : CreateCompilation(src, references: new[] { CreateCompilation(new[] { libSrc, experimentalAttributeSrc }).EmitToImageReference() });
+
+        comp.VerifyDiagnostics();
+    }
+
+    [Theory, CombinatorialData]
+    public void OnTypeParameter(bool inSource)
+    {
+        // Ignored on type parameters
+        var libSrc = """
+public class C<[System.Diagnostics.CodeAnalysis.Experimental("DiagID1")] T> { }
+""";
+
+        var src = """
+C<int> c = null;
+c.ToString();
+""";
+
+        var comp = inSource
+            ? CreateCompilation(new[] { src, libSrc, experimentalAttributeSrc })
+            : CreateCompilation(src, references: new[] { CreateCompilation(new[] { libSrc, experimentalAttributeSrc }).EmitToImageReference() });
+
+        comp.VerifyDiagnostics();
+    }
+
+    [Fact]
+    public void NullDiagnosticId()
+    {
+        var src = """
+C.M();
+
+[System.Diagnostics.CodeAnalysis.Experimental(null)]
+class C
+{
+    public static void M() { }
+}
+""";
+        var comp = CreateCompilation(new[] { src, experimentalAttributeSrc });
+        comp.VerifyDiagnostics(
+            // 0.cs(1,1): warning CS8305: 'C' is for evaluation purposes only and is subject to change or removal in future updates.
+            // C.M();
+            Diagnostic(ErrorCode.WRN_Experimental, "C").WithArguments("C").WithLocation(1, 1)
+            );
+    }
+
+    [Fact]
+    public void EmptyDiagnosticId()
+    {
+        var src = """
+C.M();
+
+[System.Diagnostics.CodeAnalysis.Experimental(null)]
```

Comment:
Did you mean an empty string here? Also consider testing a space only string, e.g., `" "`, `"\n"`.

```suggestion
[System.Diagnostics.CodeAnalysis.Experimental("")]
```